runtime/src/accounts_index.rs

```diff
@@ -1498,31 +1453,31 @@ impl<T: IndexValue> AccountsIndex<T> {
 
     /// Get an account
     /// The latest account that appears in `ancestors` or `roots` is returned.
-    pub(crate) fn get(
-        &self,
-        pubkey: &Pubkey,
+    pub(crate) fn get2<'a, RT>(
+        &'a self,
+        pubkey: &'a Pubkey,
         ancestors: Option<&Ancestors>,
         max_root: Option<Slot>,
-    ) -> AccountIndexGetResult<'_, T> {
+        user: impl for<'b> FnOnce(Option<(&'b AccountMapEntry<T>, usize)>) -> RT,
+    ) -> RT {
         let read_lock = self.account_maps[self.bin_calculator.bin_from_pubkey(pubkey)]
             .read()
             .unwrap();
-        let account = read_lock
-            .get(pubkey)
-            .map(ReadAccountMapEntry::from_account_map_entry);
-
-        match account {
-            Some(locked_entry) => {
-                drop(read_lock);
-                let slot_list = locked_entry.slot_list();
-                let found_index = self.latest_slot(ancestors, slot_list, max_root);
-                match found_index {
-                    Some(found_index) => AccountIndexGetResult::Found(locked_entry, found_index),
-                    None => AccountIndexGetResult::NotFoundOnFork,
+
+        read_lock.get_internal(pubkey, |locked_entry| {
+            let result = match locked_entry {
+                Some(locked_entry) => {
+                    let slot_list = locked_entry.slot_list();
+                    let found_index = self.latest_slot(ancestors, slot_list, max_root);
+                    match found_index {
+                        Some(found_index) => user(Some((locked_entry, found_index))),
+                        None => user(None),
+                    }
                 }
-            }
-            None => AccountIndexGetResult::Missing(read_lock),
-        }
+                None => user(None),
+            };
+            (false, result)
```

Comment:
Noting for deeper review later.

Previous implementation of `get` could return a `AccountIndexGetResult::Found` which would hold the read lock to the caller, whereas this change only clones the AccountMapEntry, meaning the backing index entry could have been modified when this function returns to the caller.

TODO: Check every caller/entrypoint of every caller to `get()` that was replaced with `get2()` for safety